### ISSUE TITLE
race in dependency resolution causes Error: Cannot read property 'value' of undefined

### ISSUE THREAD

**Issue by metamatt** [systemjs/systemjs#644]
I upgraded from SystemJS 0.16.11 to 0.18.4 and am seeing a sporadic error in our app when we run Protractor tests under a heavy load.

In my copy of the source I'm not getting useful backtraces from the minified system.js even with the sourcemap, so I told the browser to load system.src.js and the exception location is system.src.js:533:38, where `addLoadToLinkSet` is walking `load.dependencies` and one of the entries in that list is `null`.

Looking at the places that modify that `dependencies` list, it looks like a race between linkDeclarativeModule (adds the null entry), requestLoad.then AddDependencyLoad (updates the null entry with a key and value), addLoadToLinkSet (references the null entry's value). It's a race because we only sometimes hit it, and only when the browser is under pressure. We are using Bluebird (and have set window.Promise to be Bluebird); it's possible this bug doesn't happen when using the SystemJS included polyfill Promise instead of Bluebird.

Anyway, I can repro the bug in the browser without Protractor or extra load by adding a random delay before the AddDependencyLoad step:

```
diff --git a/jspm/packages/system.src.js b/jspm/packages/system.src.js
index 64058db..199590c 100644
--- a/jspm/packages/system.src.js
+++ b/jspm/packages/system.src.js
@@ -322,6 +322,18 @@ function logloads(loads) {
     );
   }

+function delayFor(delayMs) {
+   return function(arg) {
+      return new Promise(function(resolve, reject) {
+         console.warn('begin wait');
+         setTimeout(function() {
+            console.warn('end wait');
+            resolve(arg);
+         }, delayMs);
+      });
+   };
+}
+
   var anonCnt = 0;

   // 15.2.4.5
@@ -385,11 +397,15 @@ function logloads(loads) {
           loadPromises.push(
             requestLoad(loader, request, load.name, load.address)

+            .then(delayFor(Math.random() * 100))
             // 15.2.4.6.1 AddDependencyLoad (load is parentLoad)
             .then(function(depLoad) {

               // adjusted from spec to maintain dependency order
               // this is due to the System.register internal implementation needs
+              if (index < load.dependencies.length && !load.dependencies[index]) {
+                console.warn('update dep ' + index + ' of ' + load.name);
+              }
               load.dependencies[index] = {
                 key: request,
                 value: depLoad.name
@@ -530,6 +546,9 @@ function logloads(loads) {
     var loader = linkSet.loader;

     for (var i = 0, l = load.dependencies.length; i < l; i++) {
+      if (!load.dependencies[i]) {
+         console.warn('missing dependency ' + i + ' of ' + load.name);
+      }
       var name = load.dependencies[i].value;

       if (loader.modules[name])
@@ -1869,6 +1888,7 @@ SystemJSLoader.prototype.config = function(cfg) {
         module.dependencies.push(depModule);
       }
       else {
+        console.warn('add null dep index ' + module.dependencies.length + ' of ' + module.name);
         module.dependencies.push(null);
       }
```

Let me know if this is enough info to go on or if you need more.

**Comment by guybedford** [comment]
@metamatt thanks so much for these, I managed to fully replicate and post a fix in https://github.com/ModuleLoader/es6-module-loader/commit/e36fa26e64eeb41a62a8ab87c52948fef09b0512.

**Comment by guybedford** [comment]
Released in 0.18.5.

**Comment by metamatt** [comment]
Great. I sure appreciate the quick attention and fix.
